### ISSUE TITLE
Home-manager hyprctl segfault without error message, but system one work

### ISSUE THREAD

**Issue by wescande** [hyprwm/Hyprland#8693]
### Already reported ? *

- [X] I have searched the existing open and closed issues.

### Regression?

Yes

### System Info and Version

<details> <summary>System/Version info</summary>

```sh
 $> /nix/store/l2qdkbyfvij2iwn9dgj6l0kvm29xfs55-home-manager-path/bin/hyprctl systeminfo -c
zsh: segmentation fault  hyprctl systeminfo -c
```

```sh
$> /usr/bin/hyprctl systeminfo -c
Hyprland 0.45.2 built from branch v0.45.2-b at commit 12f9a0d0b93f691d4d9923716557154d74777b0a  ([gha] Nix: update inputs).
Date: 2024-11-19
Tag: v0.45.2, commits: 12f9a0d0b93f691d4d9923716557154d74777b0a
built against aquamarine 0.5.0


flags set:
debug


System Information:
System name: Linux
Node name: wescande-glaptop
Release: 6.10.11-1rodete2-amd64
Version: #1 SMP PREEMPT_DYNAMIC Debian 6.10.11-1rodete2 (2024-10-16)


GPU information: 
00:02.0 VGA compatible controller [0300]: Intel Corporation CometLake-U GT2 [UHD Graphics] [8086:9b41] (rev 02) (prog-if 00 [VGA controller])


os-release: PRETTY_NAME="Debian GNU/Linux rodete"
NAME="Debian GNU/Linux rodete"
VERSION_CODENAME=rodete
ID=debian
HOME_URL="https://go/glinux"
SUPPORT_URL="https://go/techstop"
BUG_REPORT_URL="https://go/techstop"


plugins:

======Config-Start======


Config File: /home/wescande/.config/hypr/hyprland.conf: Read Succeeded
exec-once = /nix/store/d5bki4k0jpnw5897kbnaclzmxrhdxc8c-dbus-1.14.10/bin/dbus-update-activation-environment --systemd --all && systemctl --user stop hyprland-session.target && systemctl --user start hyprland-session.target
exec-once = mako
exec-once = waybar
exec-once = input-remapper-control --command autoload --config-dir /home/wescande/.config/input-remapper-2
# exec-once = swayidle -w timeout 300 'swaylock' timeout 600 'hyprctl dispatch dpms off' resume 'hyprctl dispatch dpms on'
exec-once = swayidle -w timeout 300 'hyprlock'

#
# Please note not all available settings / options are set here.
# For a full list, see the wiki
#

# autogenerated = 1 # remove this line to remove the warning

# See https://wiki.hyprland.org/Configuring/Monitors/
monitor=,preferred,auto,1.333333
monitor=eDP-1,2880x1620,auto,auto

workspace = 1, monitor:eDP-1
workspace = 2, monitor:eDP-1
workspace = 3, monitor:eDP-1
workspace = 4, monitor:desc:Lenovo Group Limited P27h-28 V908LHD6
workspace = 5, monitor:desc:Lenovo Group Limited P27h-28 V908LHD6
workspace = 6, monitor:desc:Lenovo Group Limited P27h-28 V908LHD6
workspace = 7, monitor:desc:Lenovo Group Limited P27h-28 V908LHD6
workspace = 8, monitor:desc:Lenovo Group Limited P27h-28 V908LHD6
workspace = 9, monitor:desc:Lenovo Group Limited P27h-28 V908LHD6
workspace = 10, monitor:desc:Lenovo Group Limited P27h-28 V908LHD6
# workspace = 4, monitor:DP-4
# workspace = 5, monitor:DP-4
# workspace = 6, monitor:DP-4
# workspace = 7, monitor:DP-4
# workspace = 8, monitor:DP-4
# workspace = 9, monitor:DP-4
# workspace = 10, monitor:DP-4

# See https://wiki.hyprland.org/Configuring/Keywords/ for more

# Execute your favorite apps at launch
# exec-once = waybar & hyprpaper & firefox

# Source a file (multi-file configs)
# source = ~/.config/hypr/myColors.con=

# Set programs that you use
$terminal = kitty
$fileManager = dolphin

# Some default env vars.
env = XCURSOR_SIZE,24
env = QT_QPA_PLATFORMTHEME,qt5ct # change to qt6ct if you have that
# env = QT_AUTO_SCREEN_SCALE_FACTOR,2
env = QT_QPA_PLATFORM,wayland;xcb
env = SSH_AUTH_SOCK,$XDG_RUNTIME_DIR/openssh_agent

# For all categories, see https://wiki.hyprland.org/Configuring/Variables/
input {
    kb_layout = us
    kb_variant =
    kb_model =
    # Using Input remapper to remap capslock to Ctrl + Q
    kb_options = compose:ralt
    kb_rules =
    repeat_delay = 200
    repeat_rate = 30

    follow_mouse = 1
    scroll_factor = 0.5

    touchpad {
        natural_scroll = yes
    }

    # -1.0 to 1.0, 0 means no modification.
    sensitivity = -0.6
}

general {
    # See https://wiki.hyprland.org/Configuring/Variables/ for more
    gaps_in = 5
    gaps_out = 20
    border_size = 2
    col.active_border = rgba(33ccffee) rgba(00ff99ee) 45deg
    col.inactive_border = rgba(595959aa)

    layout = dwindle

    # Please see https://wiki.hyprland.org/Configuring/Tearing/ before you turn this on
    allow_tearing = false
}

decoration {
    # See https://wiki.hyprland.org/Configuring/Variables/ for more
    rounding = 10

    blur {
        enabled = false
        size = 3
        passes = 1
    }

    shadow {
        enabled = false
        range = 4
        render_power = 3
        color = rgba(1a1a1aee)
    }
}

animations {
    enabled = yes

    # Some default animations, see https://wiki.hyprland.org/Configuring/Animations/ for more

    bezier = overshot,      0.05,  0.9, 0.1,  1.1
    # Curb that will bounce a bit
    bezier = easeInOutBack, 0.68, -0.6, 0.32, 1.6

    animation = windows, 1, 7, overshot
    animation = windowsOut, 1, 7, default, popin 80%
    animation = border, 1, 10, default
    animation = borderangle, 1, 8, default
    animation = fade, 1, 7, default
    animation = workspaces, 1, 6, default
    # Make the special workspace come from top
    animation = specialWorkspace, 1, 6, easeInOutBack, slidefadevert -100%
}

dwindle {
    # See https://wiki.hyprland.org/Configuring/Dwindle-Layout/ for more
    pseudotile = yes # master switch for pseudotiling. Enabling is bound to mainMod + P in the keybinds section below
    preserve_split = yes # you probably want this
    force_split = 2
}

# Ref https://wiki.hyprland.org/Configuring/Workspace-Rules/
# "Smart gaps" / "No gaps when only"
workspace = s[false] w[tv1], gapsout:0, gapsin:0
workspace = s[false] f[1], gapsout:0, gapsin:0
windowrulev2 = bordersize 0, floating:0, onworkspace:s[false] w[tv1]
windowrulev2 = rounding 0, floating:0, onworkspace:s[false] w[tv1]
windowrulev2 = bordersize 0, floating:0, onworkspace:s[false] f[1]
windowrulev2 = rounding 0, floating:0, onworkspace:s[false] f[1]

# Remove border on window when there is only one tiled window in the workspace except on the special workspace
windowrulev2 = noborder, onworkspace:w[t1]r[0-9]
windowrulev2 = rounding 0, onworkspace:w[t1]r[0=9]
windowrulev2 = float, class:google-chrome, title:^Google Single Sign On.*$
windowrulev2 = float, class:firefox-esr, title:^.*Bitwarden.*$
windowrulev2 = float, title:(Extension: \(Bitwarden - Free Password Manager\) - Bitwarden — Mozilla Firefox)
windowrulev2 = float, title:(.*)(Bitwarden — Mozilla Firefox)

master {
    # See https://wiki.hyprland.org/Configuring/Master-Layout/ for more
    new_status = inherit
}

gestures {
    # See https://wiki.hyprland.org/Configuring/Variables/ for more
    workspace_swipe = on
}

misc {
    # See https://wiki.hyprland.org/Configuring/Variables/ for more
    force_default_wallpaper = 0 # Set to 0 or 1 to disable the anime mascot wallpapers
    # if there is a fullscreen or maximized window, decide whether a new tiled window opened should replace it, stay behind or disable the fullscreen/maximized state. 0 - behind, 1 - takes over, 2 - unfullscreen/unmaxize [0/1/2]
    new_window_takes_over_fullscreen = 2
}

# Example per-device config
# See https://wiki.hyprland.org/Configuring/Keywords/#per-device-input-configs for more
device {
    name = epic-mouse-v1
    sensitivity = -0.5
}
device {
    name=syna8006:00-06cb:cd8b-touchpad
    # -1.0 to 1.0, 0 means no modification.
    sensitivity = 0
}

xwayland {
    force_zero_scaling = true
}

layerrule = noanim, launcher
# layerrule = dimaround, launcher
# layerrule = blurpopups, launcher
# layerrule = blur, launcher

# Example windowrule v1
# windowrule = float, ^(kitty)$
# Example windowrule v2
# windowrulev2 = float,class:^(kitty)$,title:^(kitty)$
# See https://wiki.hyprland.org/Configuring/Window-Rules/ for more
windowrulev2 = suppressevent maximize, class:.* # You'll probably like this.

$browser=/usr/bin/flatpak run --branch=stable --arch=x86_64 --command=firefox --file-forwarding org.mozilla.firefox


# See https://wiki.hyprland.org/Configuring/Keywords/ for more
$mainMod = SUPER

# Example binds, see https://wiki.hyprland.org/Configuring/Binds/ for more
bind = $mainMod, Q, exec, $terminal
bind = $mainMod, return, exec, $terminal
bind = $mainMod, W, killactive,
bind = $mainMod, F, fullscreen,
bind = $mainMod, M, exit,
bind = $mainMod, E, exec, $fileManager
bind = $mainMod, V, togglefloating,
bindr = $mainMod, SUPER_L, exec, killall fuzzel || fuzzel
bind = $mainMod, P, pseudo, # dwindle
bind = $mainMod, F2, exec, $browser
bind = SUPER_SHIFT, F2, exec, $browser --private-window
bind = CTRL_ALT, L, exec, hyprlock

# bind = $mainMod, J, togglesplit, # dwindle

# Move focus with mainMod + arrow keys
bind = $mainMod, left, movefocus, l
bind = $mainMod, right, movefocus, r
bind = $mainMod, up, movefocus, u
bind = $mainMod, down, movefocus, d
bind = $mainMod, H, movefocus, l
bind = $mainMod, L, movefocus, r
bind = $mainMod, K, movefocus, u
bind = $mainMod, J, movefocus, d

# Switch workspaces with mainMod + [0-9]
binds {
    allow_workspace_cycles = true
}
bind = $mainMod, grave, workspace, previous

bind = $mainMod, 1, workspace, 1
bind = $mainMod, 2, workspace, 2
bind = $mainMod, 3, workspace, 3
bind = $mainMod, 4, workspace, 4
bind = $mainMod, 5, workspace, 5
bind = $mainMod, 6, workspace, 6
bind = $mainMod, 7, workspace, 7
bind = $mainMod, 8, workspace, 8
bind = $mainMod, 9, workspace, 9
bind = $mainMod, 0, workspace, 10

# Move active window to a workspace with mainMod + SHIFT + [0-9]
bind = $mainMod SHIFT, 1, movetoworkspace, 1
bind = $mainMod SHIFT, 2, movetoworkspace, 2
bind = $mainMod SHIFT, 3, movetoworkspace, 3
bind = $mainMod SHIFT, 4, movetoworkspace, 4
bind = $mainMod SHIFT, 5, movetoworkspace, 5
bind = $mainMod SHIFT, 6, movetoworkspace, 6
bind = $mainMod SHIFT, 7, movetoworkspace, 7
bind = $mainMod SHIFT, 8, movetoworkspace, 8
bind = $mainMod SHIFT, 9, movetoworkspace, 9
bind = $mainMod SHIFT, 0, movetoworkspace, 10

# Example special workspace (scratchpad)
bind = $mainMod, S, togglespecialworkspace, magic
bind = $mainMod SHIFT, S, movetoworkspace, special:magic

# Scroll through existing workspaces with mainMod + scroll
bind = $mainMod, mouse_down, workspace, e+1
bind = $mainMod, mouse_up, workspace, e-1

# Move/resize windows with mainMod + LMB/RMB and dragging
bindm = $mainMod, mouse:272, movewindow
bindm = $mainMod, mouse:273, resizewindow

bindel=, XF86AudioRaiseVolume, exec, wpctl set-volume @DEFAULT_AUDIO_SINK@ 5%+
bindel=, XF86AudioLowerVolume, exec, wpctl set-volume @DEFAULT_AUDIO_SINK@ 5%-
bindl=, XF86AudioMute, exec, wpctl set-mute @DEFAULT_AUDIO_SINK@ toggle

binde = , XF86MonBrightnessUp, exec, brightnessctl set 5%+
binde = , XF86MonBrightnessDown, exec, brightnessctl set 5%-

bindl=,switch:Lid Switch,exec,hyprlock

bind = , Print, exec, grim -g "$(slurp -d)" - | wl-copy

# bindl=,switch:on:Lid Switch,exec,bash -c 'echo "coucou" >> /home/wescande/totoLid'
# bindl=,switch:off:Lid Switch,exec,bash -c 'echo "coucou2" >> /home/wescande/totoLid'

======Config-End========



```

</details>

### Description

When running `hyprctl` command, it cause a segmentation fault.
For some reason, the only command that worked was `hyprctl instances` (the `--help` also work)

In parallel of the home-manager install, I have a installation from my system package manager (`apt install hyrpland`) the version is pretty outdated (0.41.2) but the hyprctl binary from this install is the only one that work

### How to reproduce

I am on Debian and I am trying to have a more recent Hyprland version thanks to an nix home-manager installation.

In the HM, my config  for hyprland is:
```
  wayland.windowManager.hyprland = {
    enable = true;
    package = (config.lib.nixGL.wrap pkgs.hyprland);
    systemd.variables = ["--all"];
    extraConfig = ''
''; # I didn't past this part as this is already in the systeminfo above
```

nixgl is provided by home-manager with:

<table>
  <tr>
    <th>Flake.nix</th>
    <th>home.nix</th>
  </tr>
  <tr>
    <td><pre><code>nixgl = {
url = "github:nix-community/nixGL";
inputs.nixpkgs.follows = "nixpkgs";
};</code></pre></td>
    <td><pre><code>nixGL.packages = nixgl.packages;
nixGL.defaultWrapper = "mesa";
nixGL.installScripts = [ "mesa" ];</code></pre></td>
    </tr>
</table> 

+ I have manually updated `/usr/share/wayland-sessions/hyprland.desktop` to use the binary inside the nix/store/....

### Crash reports, logs, images, videos

I am unable to find any logs related to the crash.

Please let me know how I can help investigate this more

**Comment by fufexan** [comment]
Can you run `strace hyprctl <something>`? This might provide info about what call triggers the segfault. Also debug stacktraces could help. See https://wiki.hyprland.org/Nix/Contributing-and-Debugging/#getting-a-debug-stacktrace.

**Comment by wescande** [comment]
Thanks for putting me on the good direction

I am attaching the output of `strace hyprctl version`: [strace_hyprctl_version_stderr.log](https://github.com/user-attachments/files/18117658/strace_hyprctl_version_stderr.log)

I also noticed that `/var/log/syslog` is getting populated with  this in loop
```
hyprctl[48825]: segfault at 10 ip 0000000000220da3 sp 00007ffca3b98250 error 4 in hyprctl[1f
da3,218000+31000] likely on CPU 1 (core 1, socket 0)
Code: 07 00 00 85 db 0f 88 fc 04 00 00 4c 8d 25 95 ad 02 00 49 83 7c 24 08 00 0f 84 f1 03 00
 00 e8 f4 7a ff ff 89 c7 e8 dd 79 ff ff <44> 8b 70 10 41 83 fe 09 0f 86 6c 07 00 00 44 89 f1 41 bd 01 00 00
```

>  Also debug stacktraces could help. See https://wiki.hyprland.org/Nix/Contributing-and-Debugging/#getting-a-debug-stacktrace.

I'll try later. I have never been able to build hyprland from source yet on Debian, but now that I have a nix install, I may be able to do it.

**Comment by wescande** [comment]
Ok... after trying for a few hours to have debug symbol and a gdb stack trace I gave up and end up doing the old good `print("{}", __LINE__)` .
I had a lot of issue with the documentation as I found multiple way of doing it, but in the end they either didn't work or don't include the symbol.

----

I found the issue is in `hyprctl/main.cpp`, in `request`

This line is segfaulting on my installation:
```
    const std::string USERID = std::to_string(getpwuid(getuid())->pw_uid);
```

In my case `getpwuid()` is returning null. errno is not set to any specific error message so I don't know the reason why it is null.

Note that this `getpwuid(getuid())->pw_uid` is used in a couple other place in this file.  
In `requestHyprpaper` and in `getRuntimeDir`

it looks lie `getpwuid(getuid())->pw_uid` could simply be replaced by `getuid()` (works on my machine) ?
I don't know why would the value within the getpwuid struct would be different thant the getuid, but I am not familiar with that

I could upload a pull request, but I am not at all familiar with this project (I don't even know how to build it properly) so I am happy to defer.
